platform/mbed_error.c

```diff
@@ -188,24 +266,89 @@ WEAK MBED_NORETURN mbed_error_status_t mbed_error(mbed_error_status_t error_stat
         (void) handle_error(error_status, error_value, filename, line_number, MBED_CALLER_ADDR());
 
         //On fatal errors print the error context/report
-        ERROR_REPORT(&last_error_ctx, error_msg, filename, line_number);
+        ERROR_REPORT(&last_error_ctx, error_msg, filename, line_number, true);
+    }
+    
+#if MBED_CONF_PLATFORM_CRASH_CAPTURE_ENABLED
+    uint32_t crc_val = 0;
+    crc_val = compute_crc32( (unsigned char *)report_error_ctx, ((uint32_t)&(report_error_ctx->crc_error_ctx) - (uint32_t)report_error_ctx) );
+    //Read report_error_ctx and check if CRC is correct for report_error_ctx
+    if(report_error_ctx->crc_error_ctx == crc_val) {
+        uint32_t current_reboot_count = report_error_ctx->error_reboot_count;
+        last_error_ctx.error_reboot_count = current_reboot_count + 1;
+    } else {
+        last_error_ctx.error_reboot_count = 1;
     }
+    last_error_ctx.is_error_processed = 0;//Set the flag that this is a new error
+    //Update the struct with crc
+    last_error_ctx.crc_error_ctx = compute_crc32( (unsigned char *)&last_error_ctx, ((uint32_t)&(last_error_ctx.crc_error_ctx) - (uint32_t)&last_error_ctx) );
+    memcpy(report_error_ctx, &last_error_ctx, sizeof(mbed_error_ctx));
+    //We need not call delete_mbed_crc(crc_obj) here as we are going to reset the system anyway, and calling delete while handling a fatal error may cause nested exception
+#if MBED_CONF_PLATFORM_FATAL_ERROR_AUTO_REBOOT_ENABLED
+    system_reset();//do a system reset to get the system rebooted
+    while(1);
+#endif
+#endif
 
     mbed_halt_system();
+    
+    return MBED_ERROR_FAILED_OPERATION;
 }
 
 //Register an application defined callback with error handling
 mbed_error_status_t mbed_set_error_hook(mbed_error_hook_t error_hook_in)
 {
     //register the new hook/callback
-    if (error_hook_in != NULL)  {
+    if (error_hook_in != NULL) {
         error_hook = error_hook_in;
         return MBED_SUCCESS;
     }
 
     return MBED_ERROR_INVALID_ARGUMENT;
 }
 
+//Reset the reboot error context
+mbed_error_status_t mbed_reset_reboot_error_info()
+{
+#if MBED_CONF_PLATFORM_CRASH_CAPTURE_ENABLED
+    memset(report_error_ctx, 0, sizeof(mbed_error_ctx) ); 
```

Comment:
mbed_reset_reboot_error_info - Is this thread safe?

Reply:
Its currently not, I agree that its better to make it thread-safe to avoid any assumptions on app side. I'll push a fix for this.

Reply:
I was thinking of adding in docs " Not thread safe" instead of making it thread safe, as we will have more corner cases

Reply:
Since the user cannot control when the error may get triggered I think its better to protect this struct anyway.